pkg/capi/gcp.go

```diff
@@ -0,0 +1,308 @@
+	DescribeTable("should configure Confidential VM correctly", framework.LabelCAPI, framework.LabelDisruptive,
+		func(confidentialCompute *gcpv1.ConfidentialComputePolicy) {
+			mapiProviderSpec := getGCPMAPIProviderSpec(cl)
+			Expect(mapiProviderSpec).ToNot(BeNil())
+			gcpMachineTemplate = createGCPMachineTemplate(mapiProviderSpec)
+			gcpMachineTemplate.Spec.Template.Spec.ConfidentialCompute = confidentialCompute
+			gcpMachineTemplate.Spec.Template.Spec.InstanceType = "n2d-standard-4"
+
+			if *confidentialCompute == gcpv1.ConfidentialComputePolicyEnabled {
+				mapiProviderSpec.OnHostMaintenance = OnHostMaintenanceTerminate
+				gcpMachineTemplate.Spec.Template.Spec.OnHostMaintenance = (*gcpv1.HostMaintenancePolicy)(&mapiProviderSpec.OnHostMaintenance)
+			} else {
+				mapiProviderSpec.OnHostMaintenance = "Migrate"
+				gcpMachineTemplate.Spec.Template.Spec.OnHostMaintenance = (*gcpv1.HostMaintenancePolicy)(&mapiProviderSpec.OnHostMaintenance)
+			}
+
+			if err := cl.Create(ctx, gcpMachineTemplate); err != nil && !apierrors.IsAlreadyExists(err) {
+				Expect(err).ToNot(HaveOccurred())
+			}
+
+			By("Creating a MachineSet for Confidential VM")
+			machineSet, err = framework.CreateCAPIMachineSet(ctx, cl, framework.NewCAPIMachineSetParams(
+				"gcp-machineset-confidential-74703",
+				clusterName,
+				mapiProviderSpec.Zone,
+				1,
+				corev1.ObjectReference{
+					Kind:       "GCPMachineTemplate",
+					APIVersion: infraAPIVersion,
+					Name:       gcpMachineTemplateName,
+				},
+			))
+			Expect(err).ToNot(HaveOccurred(), "Failed to create CAPI MachineSet with Confidential VM configuration")
+
+			framework.WaitForCAPIMachinesRunning(ctx, cl, machineSet.Name)
+
+			By("Verifying the Confidential VM configuration on the created GCP MachineTemplate")
+			createdTemplate := &gcpv1.GCPMachineTemplate{}
+			Expect(cl.Get(framework.GetContext(), client.ObjectKey{
+				Namespace: framework.ClusterAPINamespace,
+				Name:      gcpMachineTemplateName,
+			}, createdTemplate)).To(Succeed())
+			var confidentialComputevalue = *createdTemplate.Spec.Template.Spec.ConfidentialCompute
+			Expect(fmt.Sprintf("%v", confidentialComputevalue)).To(Equal(fmt.Sprintf("%v", *confidentialCompute)))
+		},
+		Entry("Confidential Compute enabled", &confidentialComputeEnabled),
+		Entry("Confidential Compute disabled", &confidentialComputeDisabled),
+	)
+	It("should provision Preemptible machine successfully", func() {
+		mapiProviderSpec := getGCPMAPIProviderSpec(cl)
+		Expect(mapiProviderSpec).ToNot(BeNil())
+		gcpMachineTemplate = createGCPMachineTemplate(mapiProviderSpec)
+		gcpMachineTemplate.Spec.Template.Spec.Preemptible = true
+		mapiProviderSpec.OnHostMaintenance = OnHostMaintenanceTerminate
+		gcpMachineTemplate.Spec.Template.Spec.OnHostMaintenance = (*gcpv1.HostMaintenancePolicy)(&mapiProviderSpec.OnHostMaintenance)
+
+		if err := cl.Create(ctx, gcpMachineTemplate); err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
What happens if it already exists? It might have a different configuration?

Reply:
great catch it might cause problems then .